### PR TITLE
Memory code cleanup

### PR DESCRIPTION
**Pull request by davxy**
**Memory allocator generic code cleanup.**

----

The `CRYPTO_MEM_CHECK` modes (e.g. `CRYPTO_MEM_CHECK_OFF`) are supposed to be used only by the `CRYPTO_mem_ctrl()` function.

The `CRYPTO_mem_ctrl()` function is defined only if `OPENSSL_NO_CRYPTO_MDEBUG` is not
defined, thus makes sense to define the `CRYPTO_MEM_CHECK` modes under the same condition.

**Maybe the macros can be removed at all since the definitions are not used in the overall codebase:**
1. `CRYPTO_mem_ctrl(int type)` just returns -1 and ignores the type parameter
2. `CRYPTO_mem_ctr()` has been declared DEPRECATED from 3.0

----

Convenience typedefs for user-provided CRYPTO malloc, realloc and free functions to be used around the code.

----

The `crypto_ex_data_st` structure and macros were moved near the other CRYPTO_EX stuff and functions. Just to keep related stuff close to each other.

### CODE REVIEW

File: include/openssl/crypto.h [hunk 1]
```diff
@@ -297,14 +290,16 @@ typedef struct crypto_threadid_st {
...
 
-int CRYPTO_set_mem_functions(
-        void *(*m) (size_t, const char *, int),
-        void *(*r) (void *, size_t, const char *, int),
-        void (*f) (void *, const char *, int));
-void CRYPTO_get_mem_functions(
-        void *(**m) (size_t, const char *, int),
-        void *(**r) (void *, size_t, const char *, int),
-        void (**f) (void *, const char *, int));
+typedef void *CRYPTO_malloc_func(size_t num, const char *file, int line);
```
Comment (richsalz) [inline include/openssl/crypto.h]: These typedef's need to be docmented.

Comment (davxy) [reply include/openssl/crypto.h]: Done

**Comment by davxy** [comment]
Travis-CI failure log:

```
The command "if test -n "$GENERATE" && ! $make build_all_generated; then echo -e '\052\052 FAILED -- MAKE BUILD_ALL_GENERATED'; travis_terminate 1; fi" exited with 0.

$ if test -n "$CHECKDOCS" && ! $make doc-nits; then echo -e '\052\052 FAILED -- MAKE DOC-NITS'; travis_terminate 1; fi

(cd .; /usr/bin/perl util/find-doc-nits -n -l -e )

doc/man3/OPENSSL_malloc.pod:1: CRYPTO_malloc_func is not public

doc/man3/OPENSSL_malloc.pod:1: CRYPTO_realloc_func is not public

doc/man3/OPENSSL_malloc.pod:1: CRYPTO_free_func is not public

doc/man3/OPENSSL_malloc.pod:1: *CRYPTO_malloc_func missing from NAME section

doc/man3/OPENSSL_malloc.pod:1: *CRYPTO_realloc_func missing from NAME section

doc/man3/OPENSSL_malloc.pod:1: CRYPTO_realloc_func missing from SYNOPSIS

doc/man3/OPENSSL_malloc.pod:1: CRYPTO_malloc_func missing from SYNOPSIS

Makefile:1744: recipe for target 'doc-nits' failed

make: *** [doc-nits] Error 1

** FAILED -- MAKE DOC-NITS
```

But the missing stuff which makes Travis complain is indeed there (or I've missed something about writing docs?!?)

**Comment by richsalz** [comment]
you need to add the typedef names to util/other.syms.

**Comment by iamamoose** [comment]
This pull request is ready to merge

**Comment by paulidale** [comment]
Merged to master, thanks for the contribution.

**Pull request closed without merging**